server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java

```diff
@@ -0,0 +1,205 @@
+    }
+
+    @Test
+    public void testNullArrayInnerTypeArgument() throws Exception {
+        assertNormalize("array_append([null], 1)", isLiteral(Arrays.asList(null, 1)));
+    }
+
+    @Test
+    public void testOneNullArrayInnerTypeArgumentNullObject() throws Exception {
+        assertNormalize("array_append([null, 1], null)", isLiteral(Arrays.asList(null, 1, null)));
+    }
+
+    @Test
+    public void testEmptyArrayArgument() throws Exception {
+        assertNormalize("array_append([], 1)", isLiteral(List.of(1)));
+    }
+
+    @Test
+    public void testDifferentConvertableInnerTypesLong() throws Exception {
+        assertNormalize("array_append([1::integer], 2::long)", isLiteral(Arrays.asList(1L, 2L)));
+    }
+
+    @Test
+    public void testCastTextToIntegerInnerTypes() throws Exception {
+        assertNormalize("array_append([1,2], '3')", isLiteral(Arrays.asList(1, 2, 3)));
+    }
+
+    @Test
+    public void testCastTextToDoubleInnerTypes() throws Exception {
+        assertNormalize("array_append([1,2.0], '3.0')", isLiteral(Arrays.asList(1.0, 2.0, 3.0)));
+    }
+
+    @Test
+    public void testDifferentConvertableInnerTypesDouble() throws Exception {
+        assertNormalize("array_append([1::integer], 2::double)", isLiteral(Arrays.asList(1.0, 2.0)));
+    }
+
+    @Test
+    public void test2DArrayArgumentWithArrayObject() throws Exception {
+        assertNormalize("array_append([[1,2],[3,4]], [5,6])", isLiteral(Arrays.asList(Arrays.asList(1,2), Arrays.asList(3,4), Arrays.asList(5,6))));
+    }
+
+    @Test
+    public void test2DArrayNullArgumentWithArrayObject() throws Exception {
+        assertNormalize("array_append(null, [5,6])", isLiteral(List.of(Arrays.asList(5, 6))));
+    }
+
+    @Test
+    public void test2DArrayArgumentSingleNullValueWithArrayObject() throws Exception {
+        assertNormalize("array_append([null], [5,6])", isLiteral(Arrays.asList(null, Arrays.asList(5, 6))));
+    }
+
+    @Test
+    public void test2DArrayArgumentWithNullObject() throws Exception {
+        assertNormalize("array_append([[1,2],[3,4]], null)", isLiteral(Arrays.asList(Arrays.asList(1,2), Arrays.asList(3,4), null)));
+    }
+
+    @Test
+    public void test2DArrayArgumentIncludingNullArrayWithArrayObject() throws Exception {
+        assertNormalize("array_append([[1,2],null], [5,6])", isLiteral(Arrays.asList(Arrays.asList(1,2), null, Arrays.asList(5,6))));
+    }
+
+    @Test
+    public void test2DArrayArgumentIncludingNullValueWithArrayObject() throws Exception {
+        assertNormalize("array_append([[1,2],[null, 4]], [5,6])", isLiteral(Arrays.asList(Arrays.asList(1,2), Arrays.asList(null,4), Arrays.asList(5,6))));
+    }
+
+    @Test
+    public void test2DArrayArgumentWithArrayObjectIncludingNullValue() throws Exception {
+        assertNormalize("array_append([[1,2],[3,4]], [null, 6])", isLiteral(Arrays.asList(Arrays.asList(1,2), Arrays.asList(3,4), Arrays.asList(null,6))));
+    }
```

Comment:
As the scalar implementation makes no difference if the array inner type is an array or not, I think these test case can be safely omitted.